content/2017/02/09/generating-documentation-polyglot-projects.rst

```diff
@@ -0,0 +1,124 @@
+
+The Multi-language Challenge
+----------------------------
+
+Generating documentation from inline comments like this is ideal
+for developer-oriented documentation (that is, documentation of
+how the system is built rather than how to use it). It keeps
+the documentation right next to the code that it documents,
+which both makes it more accessible and increases the chances
+that it will be kept up to date.
+
+But SassDoc only knows how to read comments from Sass.
+And the projects we work on implement patterns using multiple languages
+working together: Sass for stylesheets, Nunjucks/Jinja2 for markup,
+Javascript for interactivity, Python on the backend.
+How can we generate documentation for a pattern that involves
+multiple languages, without giving up on the goal of writing
+documentation inline?
+
+The Python documentation tool `Sphinx <http://www.sphinx-doc.org/>`_
+has a nice pattern for this: rather than only fetching documentation
+from one particular kind of code source file, it allows for
+writing free-form documentation organized in whatever way makes sense,
+but with the ability to use "autodoc" directives to pull in documentation
+from inline source comments wherever makes sense. For example, this
+directive would add documentation generated from the code and comments
+in the ``rstblog`` module:
+
+.. code:: rst
+
+  .. automodule:: rstblog
+
+But this option still suffers from the single-language problem! Sphinx's autodoc
+extension is focused on Python code. And while it is extensible,
+there is a challenge in creating good autodoc extensions for other languages:
+different languages use different syntaxes, so need to be parsed by a tool
+that understands the language. But often a high-quality parser of a particular language
+is not available in the Python ecosystem. So Sphinx autodoc extensions to pull in
+inline documentation from other languages are not consistently available
+or well-maintained.
+
+A Way Forward
+-------------
+
+In order to provide more flexibility, I propose tackling this challenge
+using a decoupled architecture: a central documentation formatter that
+parses source code using separate processes.
+
+The central formatter would work similarly to Sassdoc or Sphinx:
+read a file that specifies the overall structure of the documentation
+and look for directives that ask to include automatic documentation
+from other source files.
+
+A parsing utility would have the limited responsibility of reading
+a source file of one particular type. The central formatter would run
+the parser as a separate process and output a JSON representation of the code
+structure and comments for the central formatter to make use of.
+This way the parsing utility can be written in whatever language best
+supports parsing the source language.
+
+As a proof of concept, in the near future we intend to add a feature to
+Herman to automatically include documentation of macros from Nunjucks templates.
```

Comment:
Maybe it seems hokey, but I wonder if we could end this on a more positive or energetic note? "Stay tuned!"

Or an invitation to share ideas or thoughts with us via Slack?